File: test_pipelines.js
// Removed node-fetch require to use global fetch
const API_BASE = 'http://localhost:8081/api';

async function runTests() {
    console.log("=== Starting Data Integration API Tests ===");

    // 1. Register a fresh user or Login
    let token = '';
    const email = `testuser_${Date.now()}@example.com`;
    try {
        await fetch(`${API_BASE}/auth/register`, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({ email, password: 'password123', name: 'Test User', company_name: "Testing Co" })
        });

        const loginRes = await fetch(`${API_BASE}/auth/login`, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({ email, password: 'password123' })
        });
        const loginBody = await loginRes.text();
        const match = loginBody.match(/"token"\s*:\s*"([^"]+)"/);
        token = match ? match[1] : null;

        if (!token) throw new Error("Could not get token: " + loginBody);
        console.log("✅ Authenticated successfully");
    } catch (e) {
        console.error("❌ Authentication failed:", e.message);
        process.exit(1);
    }

    const headers = {
        'Authorization': `Bearer ${token}`,
        'Content-Type': 'application/json'
    };

    // 2. Create DataSource
    let dsId;
    try {
        const dsRes = await fetch(`${API_BASE}/datasources`, {
            method: 'POST',
            headers,
            body: JSON.stringify({
                name: "Test PostgreSQL DB",
                provider: "postgresql",
                host: "localhost",
                port: 5432,
                database: "geodata",
                username: "postgres",
                password: "1234"
            })
        });
        const dsData = await dsRes.json();
        dsId = dsData.data?.id || dsData.id;

        if (!dsId) {
            const listDsRes = await fetch(`${API_BASE}/datasources`, { headers });
            const listDsData = await listDsRes.json();
            dsId = (listDsData.data && listDsData.data[0]?.id) || (listDsData[0]?.id);
        }

        if (!dsId) throw new Error("No data source found or created: " + JSON.stringify(dsData));
        console.log("✅ Data Source Ready (ID: " + dsId + ")");
    } catch (e) {
        console.error("❌ DataSource creation failed:", e.message);
        process.exit(1);
    }

    // 3. Create Pipeline
    let pipelineId;
    try {
        const pRes = await fetch(`${API_BASE}/datasources/${dsId}/pipelines`, {
            method: 'POST',
            headers,
            body: JSON.stringify({
                name: "My Automated ETL Pipeline",
                config: {
                    base_table: "users",
                    mappings: [{ target_column: "address", expression: "address" }],
                    joins: [],
                    filters: []
                }
            })
        });
        const pData = await pRes.json();
        pipelineId = pData.data?.id || pData.id;
        if (!pipelineId) throw new Error("Pipeline creation failed: " + JSON.stringify(pData));
        console.log("✅ Pipeline created successfully (ID: " + pipelineId + ")");
    } catch (e) {
        console.error("❌ Pipeline Creation failed:", e.message);
        process.exit(1);
    }

    // 4. List Pipelines
    try {
        const listRes = await fetch(`${API_BASE}/datasources/${dsId}/pipelines`, { headers });
        const listData = await listRes.json();
        const pipelines = listData.data || listData;
        const exists = pipelines.find(p => p.id === pipelineId);
        if (!exists) throw new Error("Pipeline not found in list");
        console.log("✅ Pipeline listing works correctly.");
    } catch (e) {
        console.error("❌ Pipeline Listing failed:", e.message);
        process.exit(1);
    }

    // 5. Delete Pipeline
    try {
        const delRes = await fetch(`${API_BASE}/datasources/${dsId}/pipelines/${pipelineId}`, {
            method: 'DELETE',
            headers
        });
        if (!delRes.ok) throw new Error("Delete request failed with status: " + delRes.status);
        console.log("✅ Pipeline deleted successfully.");
    } catch (e) {
        console.error("❌ Pipeline Deletion failed:", e.message);
        process.exit(1);
    }

    console.log("=========================================");
    console.log("🚀 ALL PIPELINE END-TO-END TESTS PASSED!");
    console.log("=========================================");
}

runTests();

File: test_pipelines.py
import requests
import json
import uuid

API_BASE = 'http://localhost:8081/api'

def run():
    print("=== Starting Data Integration API Tests (Python) ===")
    
    email = f"pyuser_{uuid.uuid4()}@example.com"
    
    # 1. Register
    requests.post(f"{API_BASE}/auth/register", json={
        "email": email,
        "password": "password123",
        "name": "Py User",
        "company_name": "Py Co"
    })
    
    # 2. Login
    res = requests.post(f"{API_BASE}/auth/login", json={
        "email": email,
        "password": "password123"
    })
    data = res.json()
    token = data.get('access_token')
    if not token and 'data' in data:
        token = data['data'].get('access_token')
        
    if not token:
        print("❌ Auth failed:", data)
        return
    print("✅ Authenticated successfully")
    
    headers = {"Authorization": f"Bearer {token}"}
    
    # 3. Create DS
    ds_res = requests.post(f"{API_BASE}/datasources", headers=headers, json={
        "name": "Py Test DB",
        "provider": "postgresql",
        "host": "localhost",
        "port": 5432,
        "database": "postgres",
        "username": "postgres",
        "password": "password"
    })
    
    if ds_res.status_code == 200:
        ds_id = ds_res.json().get('id') or ds_res.json().get('data', {}).get('id')
    else:
        print("Failed to create DS:", ds_res.text)
        # fetch existing
        ds_list_res = requests.get(f"{API_BASE}/datasources", headers=headers)
        if ds_list_res.status_code != 200:
            print("Failed to fetch DS:", ds_list_res.text)
            return
            
        ds_list = ds_list_res.json()
        if 'data' in ds_list:
            ds_list = ds_list['data']
        ds_id = ds_list[0]['id'] if ds_list and len(ds_list) > 0 else None
        
    if not ds_id:
        print("❌ Could not get or create DataSource")
        return
        
    print(f"✅ Data Source Ready (ID: {ds_id})")
    
    # 4. Create Pipeline
    print("\n--- Creating Pipeline ---")
    p_res = requests.post(f"{API_BASE}/pipelines", headers=headers, json={
        "name": "My Automates ETL",
        "data_source_id": ds_id,
        "config": {
            "base_table": "public.orders",
            "joins": [],
            "mappings": [
                {"target_column": "full_address", "expression": "delivery_address"}
            ],
            "cron_active": True,
            "cron": "*/1 * * * *" # Every minute
        }
    })
    print("Create Pipeline Status:", p_res.status_code)
    if p_res.status_code == 200:
        p_data = p_res.json()
        p_id = p_data.get('id') or p_data.get('data', {}).get('id')
        print("Pipeline Created:", p_data)
    else:
        print("Create fail:", p_res.text)
        return
        
    print(f"✅ Pipeline created successfully (ID: {p_id})")
    
    # 5. List Pipelines
    list_res = requests.get(f"{API_BASE}/datasources/{ds_id}/pipelines", headers=headers)
    pipelines = list_res.json()
    if 'data' in pipelines:
        pipelines = pipelines['data']
        
    found = any(p.get('id') == p_id for p in pipelines)
    if not found:
        print("❌ Pipeline not found in list")
        return
    print("✅ Pipeline listing works correctly.")
    
    # 6. Delete Pipeline
    del_res = requests.delete(f"{API_BASE}/pipelines/{p_id}", headers=headers)
    if del_res.status_code != 200:
        print("❌ Pipeline Deletion failed:", del_res.text)
        return
    print("✅ Pipeline deleted successfully.")
    
    print("=========================================")
    print("🚀 ALL PIPELINE END-TO-END TESTS PASSED!")
    print("=========================================")

if __name__ == '__main__':
    run()
